rfcs/027-pipeline-intermediate-storage/README.md

```diff
@@ -0,0 +1,147 @@
+
+* There will be a performance penalty to making two sequential read or write requests to DynamoDB and S3 which other stores will not suffer from. Other databases will also make use of a cache to improve read performance, whilst the DynamoDB component of VHS will not be able to cache the data itself.
+
+* Although the VHS is used throughout our code and has proven very reliable, it still feels like a workaround to the limitations of DynamoDB which would not be present in other stores.
+
+There has already been some discussion about usage of a different store to VHS, and due to the work on the enricher requiring a new intermediate store we have an opportunity to try out an alternative. If we can improve on these issues above, in the future we could potentially backport this to other parts of the pipeline.
+
+## Requirements for Intermediate Store
+
+Taking into account the business logic of our services, and given the issues with the current VHS highlighted in [Problem](#problem), there are a number of requirements and desirable features for any alternative store:
+
+1. The ability to insert items into the store with some unique string ID and some specific integer version (set externally to the pipeline from our adapters), with a method to only store data when the version given is greater or equal to what is currently in the store.
+
+2. The ability to retrieve data from the store, where the latest data is always retrieved and the version is returned with the data.
+
+3. Happily stores data in our current range of around 8GB, and scaling beyond this for future proofing.
+
+4. Some form of disk persistence so that we do not need to perform a reindex from scratch in the case of the database going down.
+
+5. Cheaper overall cost than currently, including read / write requests, associated networks costs, and storage costs.
+
+6. Ideally improved performance.
+
+7. Ideally better visibility into the data.
+
+## Store Options
+
+There are a number of different options to use as an intermediate store that I have considered, outlined below.
+
+### Elasticsearch
+
+Pros:
+
+* Used elsewhere in the pipeline so general familiarity with it.
+
+* Kibana would provide a level of visibility into the output at different stages unlike anything else available, and the ability to perform complex queries on our data would likely be an asset to debugging.
+
+* Can access via AWS network address to reduce NAT gateway costs (see [here](https://www.elastic.co/guide/en/cloud/current/ec-getting-started-private.html)).
+
+* Back of the envelope calculations show likely to be much cheaper than current solution.
+
+* Horizontal scalability means quite capable of supporting data at sizes much beyond what we have now.
+
+Cons:
+
+* Requires use of external cloud platform or maintenance of our own Elasticsearch cluster, rather than being a service within AWS.
```

Comment:
I wonder if we would consider the AWS Elasticsearch service to mitigate this.

Reply:
Oh yeh I did know they offered elasticsearch service but slipped my mind.. Wondering whether it would still be easy to set up with Kibana etc? What were the original reasons for going with Elastic Cloud rather than this?

Reply:
AIUI, we use hosted Elasticsearch because we don't want to deal with running our own cluster – an ELK stack is a pain to maintain and gets fiddly in all sorts of ways. Paying Elastic means they deal with weird cluster issues for us.

Might be fine for internal stuff; these clusters would be more transient and if we had any issues, we could delete the cluster and restart.

Reply:
+1 to self-managed transient stores

Reply:
From Slack: 
> On reflection introducing AWS Elasticsearch may introduce a whole new bunch of failure modes we are unfamiliar with.

I am for starting with Elasticcloud ES and then considering other options if we run into issues.